api/src/task/opportunities/export_opportunity_data_task.py

```diff
@@ -0,0 +1,108 @@
+    help="Generate JSON and CSV files containing an export of all opportunity data",
+)
+@flask_db.with_db_session()
+def export_opportunity_data(db_session: db.Session) -> None:
+    ExportOpportunityDataTask(db_session).run()
+
+
+class ExportOpportunityDataTask(Task):
+    class Metrics(StrEnum):
+        RECORDS_LOADED = "records_loaded"
+
+    def __init__(
+        self,
+        db_session: db.Session,
+    ) -> None:
+        super().__init__(db_session)
+
+        FILE_NAME: str = "opportunity_data"
+        self.current_timestamp = get_now_us_eastern_datetime().strftime("%Y-%m-%d_%H-%M-%S")
+
+        # Surely there is a better way to do paths in python?
+        # I tried pathlib's Path.cwd() thinking it would resolve
+        # to /api/src/task/opportunities/ but it resolved to just /api
+        self.FILE_PATH = "/api/src/task/opportunities/output/"
+
+        self.json_file = f"{FILE_NAME}-{self.current_timestamp}.json"
+        self.csv_file = f"{FILE_NAME}-{self.current_timestamp}.csv"
+
+        self.set_metrics({"csv_file": self.csv_file, "json_file": self.json_file})
+
+    def run_task(self) -> None:
+        # load records
+        schema = OpportunityV1Schema()
+        data_to_export: dict = {
+            "metadata": {"file_generated_at": self.current_timestamp},
+            "opportunities": [],
+        }
+
+        for opp_batch in self.fetch_opportunities():
+            for record in opp_batch:
+                self.increment(self.Metrics.RECORDS_LOADED)
+                data_to_export["opportunities"].append(schema.dump(record))
+
+        # Export the data to json
+        self.export_data_to_json(data_to_export=data_to_export)
+
+        # Export the opportunities to a csv
+        self.export_opportunities_to_csv(opportunities=data_to_export["opportunities"])
+
+    def fetch_opportunities(self) -> Iterator[Sequence[Opportunity]]:
+        """
+        Fetch the opportunities in batches. The iterator returned
+        will give you each individual batch to be processed.
+
+        Fetches all opportunities where:
+            * is_draft = False
+            * current_opportunity_summary is not None
+        """
+        return (
+            self.db_session.execute(
+                select(Opportunity)
+                .join(CurrentOpportunitySummary)
+                .where(
+                    Opportunity.is_draft.is_(False),
+                    CurrentOpportunitySummary.opportunity_status.isnot(None),
+                )
+                .options(selectinload("*"), noload(Opportunity.all_opportunity_summaries))
+                .execution_options(yield_per=5000)
+            )
+            .scalars()
+            .partitions()
+        )
+
+    def export_data_to_json(self, data_to_export: dict) -> None:
+        # create the json file
+        json_object = json.dumps(data_to_export, indent=4)
+        json_file = self.FILE_PATH + self.json_file
+        with open(json_file, "w") as outfile:
```

Comment:
Rather than use open, use our file utils open_stream function. Should behave the same, but under the hood it uses the smart open library which can handle writing outputs locally or to S3 without needing to do anything special.

You can see some example usage in https://github.com/navapbc/simpler-grants-gov/blob/main/api/src/task/opportunities/import_opportunity_csvs.py#L52 - although you'll want to open it like you did here with the mode being `w`

Reply:
This should be implemented with the latest commit